cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+             for example: log_output=True, var_ordering=3, num_cores=8, ...>
+            <Add link to documentation of all solver parameters>
+        """
+        from gurobipy import GRB
+
+        if time_limit is not None:
+            raise NotImplementedError("TEMPLATE: TODO, implement time_limit")
+
+        # call the solver, with parameters
+        for param, val in kwargs.items():
+            self.gbi_model.setParam(param, val)
+
+        _ = self.gbi_model.optimize(callback=solution_callback)
+
+        my_status = self.gbi_model.Status
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.gbi_model.runtime
+
+        # translate exit status
+        if my_status == GRB.OPTIMAL and self._objective_value is None:
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status == GRB.OPTIMAL and self._objective_value is not None:
+            self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+        elif my_status == GRB.INFEASIBLE:
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        else:  # another?
+            raise NotImplementedError(
+                f"Translation of gurobi status {my_status} to CPMpy status not implemented")  # a new status type was introduced, please report on github
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return(self.cpm_status)
+
+        # translate solution values (of user vars only)
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                solver_val = self.solver_var(cpm_var).X
+                if cpm_var.is_bool():
+                    cpm_var._value = bool(solver_val)
+                else:
+                    cpm_var._value = int(solver_val)
+
+        return has_sol
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        from gurobipy import GRB
+
+        if is_num(cpm_var):
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return 1-self.solver_var(cpm_var._bv)
+
+        # create if it does not exit
+        if not cpm_var in self._varmap:
+            if isinstance(cpm_var, _BoolVarImpl):
+                revar = self.gbi_model.addVar(vtype=GRB.BINARY, name=cpm_var.name)
+            elif isinstance(cpm_var, _IntVarImpl):
+                revar = self.gbi_model.addVar(cpm_var.lb, cpm_var.ub, vtype=GRB.INTEGER, name=str(cpm_var))
+            else:
+                raise NotImplementedError("Not a known var {}".format(cpm_var))
+            self._varmap[cpm_var] = revar
+            # Update model to make new vars visible for constraints
+            self.gbi_model.update()
```

Comment:
do we really need this? I think it will cause a performance penalty for large models...

we don't introspect the gbi_model, I don't think we need to tell it to update its internal datastructures now. We can just post constraints with that variable, and it iwll 'update' the model when we ask it to solve it?

Reply:
This is a leftover artifact of a previous version I had, will remove it.